data/src/test/java/com/linkedin/data/TestDataComplexTable.java

```diff
@@ -0,0 +1,111 @@
+package com.linkedin.data;
+
+import java.util.Collection;
+import java.util.Map;
+import org.junit.Assert;
+import org.junit.Test;
+
+
+public class TestDataComplexTable
+{
+  @Test
+  public void testKeysWithSameHash() throws Exception
+  {
+    DataComplexTable table = new DataComplexTable();
+    MockDataComplex item1 = new MockDataComplex();
+    item1.setHashCode(1);
+    DataComplex item1Clone = item1.clone();
+    MockDataComplex item2 = new MockDataComplex();
+    item2.setHashCode(1);
+    DataComplex item2Clone = item2.clone();
+
+    table.put(item1, item1Clone);
+    table.put(item2, item2Clone);
+
+    Assert.assertNotNull(table.get(item1));
+    Assert.assertNotNull(table.get(item2));
+    Assert.assertSame(item1Clone, table.get(item1));
+    Assert.assertSame(item2Clone, table.get(item2));
+  }
+
+  private static class MockDataComplex implements DataComplex
```

Comment:
It might be clearer to throw `UnsupportedOperationException` in methods that not intended to be overridden.